### PR TITLE
Use DifferentiationInterface

### PR DESCRIPTION
**Pull request by avik-pal**
1. Removes AD backend-specific code (except rrules)
2. Merge https://github.com/SciML/SimpleNonlinearSolve.jl/pull/147 first
3. `SimpleHalley` now supports in place problems
4. More backends are now supported by default

TODOs

- [x] Update Halley tests to add the in place versions
- [x] Add Halley to 23 test problems

### CODE REVIEW

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by gdalle** [comment]
Any issues with DI integration @avik-pal ?

File: src/ad.jl [hunk 1]
```diff
@@ -116,43 +101,40 @@ function __nlsolve_ad(prob::NonlinearLeastSquaresProblem, alg, args...; kwargs..
...
             _F = @closure (du, u, p) -> begin
-                resid = similar(du, length(sol.resid))
-                res = DiffResults.DiffResult(
-                    resid, similar(du, length(sol.resid), length(u)))
                 _f = @closure (du, u) -> prob.f(du, u, p)
-                ForwardDiff.jacobian!(res, _f, resid, u)
-                mul!(reshape(du, 1, :), vec(DiffResults.value(res))',
-                    DiffResults.jacobian(res), 2, false)
+                resid = __similar(du, length(sol.resid))
+                v, J = DI.value_and_jacobian(_f, resid, AutoForwardDiff(), u)
```
Comment (gdalle) [inline src/ad.jl:106]: you may want to use preparation

File: src/ad.jl [hunk 2]
```diff
@@ -116,43 +101,40 @@ function __nlsolve_ad(prob::NonlinearLeastSquaresProblem, alg, args...; kwargs..
...
                 _F = @closure (u, p) -> begin
-                    T = promote_type(eltype(u), eltype(p))
-                    res = DiffResults.DiffResult(similar(u, T, size(sol.resid)),
-                        similar(u, T, length(sol.resid), length(u)))
-                    ForwardDiff.jacobian!(res, Base.Fix2(prob.f, p), u)
-                    return reshape(
-                        2 .* vec(DiffResults.value(res))' * DiffResults.jacobian(res),
-                        size(u))
+                    _f = Base.Fix2(prob.f, p)
+                    v, J = DI.value_and_jacobian(_f, AutoForwardDiff(), u)
```
Comment (gdalle) [inline src/ad.jl:121]: preparation?

File: src/ad.jl [hunk 3]
```diff
@@ -182,16 +164,12 @@ end
 
 @inline function __nlsolve_∂f_∂u(prob, f::F, u, p) where {F}
     if isinplace(prob)
-        du = similar(u)
-        __f = (du, u) -> f(du, u, p)
-        ForwardDiff.jacobian(__f, du, u)
+        __f = @closure (du, u) -> f(du, u, p)
+        return ForwardDiff.jacobian(__f, __similar(u), u)
```
Comment (gdalle) [inline src/ad.jl:168]: why not use DI?

File: src/ad.jl [hunk 4]
```diff
@@ -182,16 +164,12 @@ end
...
     else
         __f = Base.Fix2(f, p)
-        if u isa Number
-            return ForwardDiff.derivative(__f, u)
-        else
-            return ForwardDiff.jacobian(__f, u)
-        end
+        u isa Number && return ForwardDiff.derivative(__f, u)
+        return ForwardDiff.jacobian(__f, u)
```
Comment (gdalle) [inline src/ad.jl:172]: DI?

File: src/utils.jl [hunk 5]
```diff
@@ -26,209 +21,70 @@ Return the maximum of `a` and `b` if `x1 > x0`, otherwise return the minimum.
...
+function __fixed_parameter_function(prob::AbstractNonlinearProblem)
+    isinplace(prob) && return @closure (du, u) -> prob.f(du, u, prob.p)
+    return Base.Fix2(prob.f, prob.p)
 end
 
-"""
-    value_and_jacobian(ad, f, y, x, p, cache; J = nothing)
+function value_and_jacobian(
+        ad, prob::AbstractNonlinearProblem, f::F, y, x, cache; J = nothing) where {F}
+    x isa Number && return DI.value_and_derivative(f, ad, x, cache)
```
Comment (gdalle) [inline src/utils.jl:31]: was the `cache` obtained with `DI.prepare_derivative(f, ad, x)`?

File: src/utils.jl [hunk 6]
```diff
@@ -26,209 +21,70 @@ Return the maximum of `a` and `b` if `x1 > x0`, otherwise return the minimum.
...
-            return _f(x), J_fd
-        elseif ad isa AutoPolyesterForwardDiff
-            __polyester_forwarddiff_jacobian!(_f, J, x, cache)
-            return _f(x), J
-        else
-            throw(ArgumentError("Unsupported AD method: $(ad)"))
-        end
+        cache isa HasAnalyticJacobian && return f(x), prob.f.jac(x, prob.p)
+        J === nothing && return DI.value_and_jacobian(f, ad, x, cache)
+        y, J = DI.value_and_jacobian!(f, J, ad, x, cache)
```
Comment (gdalle) [inline src/utils.jl:43]: was the `cache` obtained with `DI.prepare_jacobian(f, ad, x)`?

File: src/utils.jl [hunk 7]
```diff
@@ -26,209 +21,70 @@ Return the maximum of `a` and `b` if `x1 > x0`, otherwise return the minimum.
...
-        elseif ad isa AutoFiniteDiff
-            return nothing, FiniteDiff.JacobianCache(copy(x), copy(y), copy(y), ad.fdtype)
-        else
-            throw(ArgumentError("Unsupported AD method: $(ad)"))
-        end
+function compute_jacobian_and_hessian(
+        ad, prob::AbstractNonlinearProblem, f::F, y, x) where {F}
+    if x isa Number
+        H = DI.second_derivative(f, ad, x)
+        v, J = DI.value_and_derivative(f, ad, x)
```
Comment (gdalle) [inline src/utils.jl:66]: preparation?

File: src/utils.jl [hunk 8]
```diff
@@ -26,209 +21,70 @@ Return the maximum of `a` and `b` if `x1 > x0`, otherwise return the minimum.
...
-    else
-        f = Base.Fix2(prob.f, prob.p)
-        fx = f(x)
-        J_fn = Base.Fix1(ForwardDiff.jacobian, f)
-        dfx = J_fn(x)
-        d2fx = ForwardDiff.jacobian(J_fn, x)
-        return fx, dfx, d2fx
+        df = @closure x -> begin
+            res = __similar(y, promote_type(eltype(y), eltype(x)))
+            return DI.jacobian(f, res, ad, x)
```
Comment (gdalle) [inline src/utils.jl:73]: preparation?

File: src/utils.jl [hunk 9]
```diff
@@ -26,209 +21,70 @@ Return the maximum of `a` and `b` if `x1 > x0`, otherwise return the minimum.
...
+        end
+        J, H = DI.value_and_jacobian(df, ad, x)
```
Comment (gdalle) [inline src/utils.jl:75]: preparation?

File: src/utils.jl [hunk 10]
```diff
@@ -26,209 +21,70 @@ Return the maximum of `a` and `b` if `x1 > x0`, otherwise return the minimum.
...
     if isinplace(prob)
-        error("Inplace version for Nested ForwardDiff Not Implemented Yet!")
-    else
-        f = Base.Fix2(prob.f, prob.p)
-        fx = f(x)
-        J_fn = Base.Fix1(ForwardDiff.jacobian, f)
-        dfx = J_fn(x)
-        d2fx = ForwardDiff.jacobian(J_fn, x)
-        return fx, dfx, d2fx
+        df = @closure x -> begin
```
Comment (gdalle) [inline src/utils.jl:71]: you can use DI with functions `f!(y, x)` without the closure

File: src/utils.jl [hunk 11]
```diff
@@ -360,10 +216,19 @@ end
...
+    return AutoPolyesterForwardDiff(;
+        chunksize = ForwardDiff.pickchunksize(length(prob.u0)), ad.tag)
+end
 @inline function __get_concrete_autodiff(prob, ::Nothing; kwargs...)
-    return ifelse(
-        ForwardDiff.can_dual(eltype(prob.u0)), AutoForwardDiff(), AutoFiniteDiff())
+    return ifelse(ForwardDiff.can_dual(eltype(prob.u0)),
+        AutoForwardDiff(; chunksize = ForwardDiff.pickchunksize(length(prob.u0))),
+        AutoFiniteDiff())
 end
```
Comment (gdalle) [inline src/utils.jl:232]: is this necessary with DI?

**Review by gdalle — COMMENTED** [review]
So cool that you managed to merge this! Here are a few remarks on DI use, feel free to take them into account or not

**Pull request merged**